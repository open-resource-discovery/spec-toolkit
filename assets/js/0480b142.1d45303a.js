"use strict";(self.webpackChunk_open_resource_discovery_spec_toolkit_website=self.webpackChunk_open_resource_discovery_spec_toolkit_website||[]).push([[8070],{7208:(e,t,s)=>{s.r(t),s.d(t,{assets:()=>c,contentTitle:()=>n,default:()=>h,frontMatter:()=>r,metadata:()=>i,toc:()=>d});const i=JSON.parse('{"id":"faq","title":"FAQ","description":"Q: Should I write the spec in YAML or JSON","source":"@site/docs/faq.md","sourceDirName":".","slug":"/faq","permalink":"/spec-toolkit/faq","draft":false,"unlisted":false,"tags":[],"version":"current","sidebarPosition":22,"frontMatter":{"title":"FAQ","sidebar_position":22},"sidebar":"defaultSidebar","previous":{"title":"Best Practices","permalink":"/spec-toolkit/best-practices"}}');var a=s(4848),o=s(8453);const r={title:"FAQ",sidebar_position:22},n="FAQ",c={},d=[{value:"Q: Should I write the spec in YAML or JSON",id:"q-should-i-write-the-spec-in-yaml-or-json",level:4},{value:"Q: Do you have a library for language XYZ?",id:"q-do-you-have-a-library-for-language-xyz",level:4}];function l(e){const t={a:"a",h1:"h1",h4:"h4",header:"header",p:"p",strong:"strong",...(0,o.R)(),...e.components};return(0,a.jsxs)(a.Fragment,{children:[(0,a.jsx)(t.header,{children:(0,a.jsx)(t.h1,{id:"faq",children:"FAQ"})}),"\n",(0,a.jsx)(t.h4,{id:"q-should-i-write-the-spec-in-yaml-or-json",children:"Q: Should I write the spec in YAML or JSON"}),"\n",(0,a.jsxs)(t.p,{children:["Both is possible. YAML has some ",(0,a.jsx)(t.a,{href:"https://sidneyliebrand.medium.com/the-greatnesses-and-gotchas-of-yaml-5e3377ef0c55",children:"quirks that you need to be aware of"}),"."]}),"\n",(0,a.jsxs)(t.p,{children:["The advantage over JSON is that it's easy to add comments and especially to have information ",(0,a.jsx)(t.strong,{children:"reuse"})," through YAML anchors for merging data.\nWhen writing the spec, it's very important to not duplicate information where possible, to avoid inconsistencies."]}),"\n",(0,a.jsx)(t.h4,{id:"q-do-you-have-a-library-for-language-xyz",children:"Q: Do you have a library for language XYZ?"}),"\n",(0,a.jsx)(t.p,{children:"Spec JSON Schema itself is described as standard JSON Schema and therefore the entire JSON Schema ecosystem can be used.\nOf course the Spec-Toolkit specific JSON Schema additions will not be supported by general purpose libraries.\nThe specs written with the spec-toolkit are also exported as regular JSON Schema."}),"\n",(0,a.jsxs)(t.p,{children:["Consider using standard JSON Schema validators and converter tools like ",(0,a.jsx)(t.a,{href:"https://quicktype.io/",children:"quicktype"}),".\nVSCode and other IDEs also provide validation and code intelligence by pointing to a JSON Schema file in the root of a JSON document.\nFor YAML, the ",(0,a.jsx)(t.a,{href:"https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml",children:"Red Hat YAML"})," extension will be necessary."]})]})}function h(e={}){const{wrapper:t}={...(0,o.R)(),...e.components};return t?(0,a.jsx)(t,{...e,children:(0,a.jsx)(l,{...e})}):l(e)}}}]);