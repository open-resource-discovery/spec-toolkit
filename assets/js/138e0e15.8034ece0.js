"use strict";(self.webpackChunk_open_resource_discovery_spec_toolkit_website=self.webpackChunk_open_resource_discovery_spec_toolkit_website||[]).push([[4921],{1597:e=>{e.exports=JSON.parse('{"name":"@easyops-cn/docusaurus-search-local","id":"default"}')}}]);