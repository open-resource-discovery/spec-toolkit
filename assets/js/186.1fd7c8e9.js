"use strict";(self.webpackChunk_open_resource_discovery_spec_toolkit_website=self.webpackChunk_open_resource_discovery_spec_toolkit_website||[]).push([[186],{186:(e,r,t)=>{t.d(r,{diagram:()=>a});var s=t(2030),o=(t(3222),t(6044),t(6157),t(7231),t(4007),t(2355),t(6503),t(4431),t(5395),t(2107)),a={parser:s.Zk,get db(){return new s.u4(2)},renderer:s.q7,styles:s.tM,init:(0,o.K2)((e=>{e.state||(e.state={}),e.state.arrowMarkerAbsolute=e.arrowMarkerAbsolute}),"init")}}}]);