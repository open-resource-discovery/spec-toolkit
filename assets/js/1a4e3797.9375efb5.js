"use strict";(self.webpackChunk_open_resource_discovery_spec_toolkit_website=self.webpackChunk_open_resource_discovery_spec_toolkit_website||[]).push([[2138],{1283:(e,t,r)=>{r.r(t),r.d(t,{default:()=>H});var s=r(6540),a=r(4586),n=r(3823),c=r(5260),o=r(8774),l=r(1312),u=r(5846),i=r(4164),h=r(6347),d=r(2303),m=r(1088);const p=function(){const e=(0,d.A)(),t=(0,h.W6)(),r=(0,h.zy)(),{siteConfig:{baseUrl:s}}=(0,a.A)(),n=e?new URLSearchParams(r.search):null,c=n?.get("q")||"",o=n?.get("ctx")||"",l=n?.get("version")||"",u=e=>{const t=new URLSearchParams(r.search);return e?t.set("q",e):t.delete("q"),t};return{searchValue:c,searchContext:o&&Array.isArray(m.Hg)&&m.Hg.some((e=>"string"==typeof e?e===o:e.path===o))?o:"",searchVersion:l,updateSearchPath:e=>{const r=u(e);t.replace({search:r.toString()})},updateSearchContext:e=>{const s=new URLSearchParams(r.search);s.set("ctx",e),t.replace({search:s.toString()})},generateSearchPageLink:e=>{const t=u(e);return`${s}search?${t.toString()}`}}};var g=r(9153),f=r(9913),x=r(6841),y=r(3810),j=r(7674),S=r(2849),_=r(4471);const w="searchContextInput_mXoe",v="searchQueryInput_CFBF",A="searchResultItem_U687",C="searchResultItemPath_uIbk",b="searchResultItemSummary_oZHr",P="searchQueryColumn_q7nx",F="searchContextColumn_oWAF";var R=r(3385),T=r(4848);function $(){const{siteConfig:{baseUrl:e},i18n:{currentLocale:t}}=(0,a.A)(),{selectMessage:r}=(0,u.W)(),{searchValue:n,searchContext:o,searchVersion:h,updateSearchPath:d,updateSearchContext:f}=p(),[x,y]=(0,s.useState)(n),[j,_]=(0,s.useState)(),A=`${e}${h}`,C=(0,s.useMemo)((()=>x?(0,l.T)({id:"theme.SearchPage.existingResultsTitle",message:'Search results for "{query}"',description:"The search page title for non-empty query"},{query:x}):(0,l.T)({id:"theme.SearchPage.emptyResultsTitle",message:"Search the documentation",description:"The search page title for empty query"})),[x]);(0,s.useEffect)((()=>{d(x),x?(async()=>{const e=await(0,g.w)(A,o,x,100);_(e)})():_(void 0)}),[x,A,o]);const b=(0,s.useCallback)((e=>{y(e.target.value)}),[]);(0,s.useEffect)((()=>{n&&n!==x&&y(n)}),[n]);const[$,H]=(0,s.useState)(!1);return(0,s.useEffect)((()=>{!async function(){(!Array.isArray(m.Hg)||o||m.dz)&&await(0,g.k)(A,o),H(!0)}()}),[o,A]),(0,T.jsxs)(s.Fragment,{children:[(0,T.jsxs)(c.A,{children:[(0,T.jsx)("meta",{property:"robots",content:"noindex, follow"}),(0,T.jsx)("title",{children:C})]}),(0,T.jsxs)("div",{className:"container margin-vert--lg",children:[(0,T.jsx)("h1",{children:C}),(0,T.jsxs)("div",{className:"row",children:[(0,T.jsx)("div",{className:(0,i.A)("col",{[P]:Array.isArray(m.Hg),"col--9":Array.isArray(m.Hg),"col--12":!Array.isArray(m.Hg)}),children:(0,T.jsx)("input",{type:"search",name:"q",className:v,"aria-label":"Search",onChange:b,value:x,autoComplete:"off",autoFocus:!0})}),Array.isArray(m.Hg)?(0,T.jsx)("div",{className:(0,i.A)("col","col--3","padding-left--none",F),children:(0,T.jsxs)("select",{name:"search-context",className:w,id:"context-selector",value:o,onChange:e=>f(e.target.value),children:[m.dz&&(0,T.jsx)("option",{value:"",children:(0,l.T)({id:"theme.SearchPage.searchContext.everywhere",message:"Everywhere"})}),m.Hg.map((e=>{const{label:r,path:s}=(0,R.p)(e,t);return(0,T.jsx)("option",{value:s,children:r},s)}))]})}):null]}),!$&&x&&(0,T.jsx)("div",{children:(0,T.jsx)(S.A,{})}),j&&(j.length>0?(0,T.jsx)("p",{children:r(j.length,(0,l.T)({id:"theme.SearchPage.documentsFound.plurals",message:"1 document found|{count} documents found",description:'Pluralized label for "{count} documents found". Use as much plural forms (separated by "|") as your language support (see https://www.unicode.org/cldr/cldr-aux/charts/34/supplemental/language_plural_rules.html)'},{count:j.length}))}):(0,T.jsx)("p",{children:(0,l.T)({id:"theme.SearchPage.noResultsText",message:"No documents were found",description:"The paragraph for empty search result"})})),(0,T.jsx)("section",{children:j&&j.map((e=>(0,T.jsx)(k,{searchResult:e},e.document.i)))})]})]})}function k({searchResult:{document:e,type:t,page:r,tokens:s,metadata:a}}){const n=t===f.i.Title,c=t===f.i.Keywords,l=t===f.i.Description,u=l||c,i=n||u,h=t===f.i.Content,d=(n?e.b:r.b).slice(),p=h||u?e.s:e.t;i||d.push(r.t);let g="";if(m.CU&&s.length>0){const e=new URLSearchParams;for(const t of s)e.append("_highlight",t);g=`?${e.toString()}`}return(0,T.jsxs)("article",{className:A,children:[(0,T.jsx)("h2",{children:(0,T.jsx)(o.A,{to:e.u+g+(e.h||""),dangerouslySetInnerHTML:{__html:h||u?(0,x.Z)(p,s):(0,y.C)(p,(0,j.g)(a,"t"),s,100)}})}),d.length>0&&(0,T.jsx)("p",{className:C,children:(0,_.$)(d)}),(h||l)&&(0,T.jsx)("p",{className:b,dangerouslySetInnerHTML:{__html:(0,y.C)(e.t,(0,j.g)(a,"t"),s,100)}})]})}const H=function(){return(0,T.jsx)(n.A,{children:(0,T.jsx)($,{})})}},5846:(e,t,r)=>{r.d(t,{W:()=>u});var s=r(6540),a=r(4586);const n=["zero","one","two","few","many","other"];function c(e){return n.filter((t=>e.includes(t)))}const o={locale:"en",pluralForms:c(["one","other"]),select:e=>1===e?"one":"other"};function l(){const{i18n:{currentLocale:e}}=(0,a.A)();return(0,s.useMemo)((()=>{try{return function(e){const t=new Intl.PluralRules(e);return{locale:e,pluralForms:c(t.resolvedOptions().pluralCategories),select:e=>t.select(e)}}(e)}catch(t){return console.error(`Failed to use Intl.PluralRules for locale "${e}".\nDocusaurus will fallback to the default (English) implementation.\nError: ${t.message}\n`),o}}),[e])}function u(){const e=l();return{selectMessage:(t,r)=>function(e,t,r){const s=e.split("|");if(1===s.length)return s[0];s.length>r.pluralForms.length&&console.error(`For locale=${r.locale}, a maximum of ${r.pluralForms.length} plural forms are expected (${r.pluralForms.join(",")}), but the message contains ${s.length}: ${e}`);const a=r.select(t),n=r.pluralForms.indexOf(a);return s[Math.min(n,s.length-1)]}(r,t,e)}}}}]);