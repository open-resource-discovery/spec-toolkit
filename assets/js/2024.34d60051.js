"use strict";(self.webpackChunk_open_resource_discovery_spec_toolkit_website=self.webpackChunk_open_resource_discovery_spec_toolkit_website||[]).push([[2024],{2024:(e,s,c)=>{c.d(s,{createArchitectureServices:()=>r.S});var r=c(8815);c(3579)}}]);