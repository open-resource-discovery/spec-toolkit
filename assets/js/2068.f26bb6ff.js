"use strict";(self.webpackChunk_open_resource_discovery_spec_toolkit_website=self.webpackChunk_open_resource_discovery_spec_toolkit_website||[]).push([[2068],{2068:(e,s,_)=>{_.d(s,{createInfoServices:()=>c.v});var c=_(1366);_(8198)}}]);