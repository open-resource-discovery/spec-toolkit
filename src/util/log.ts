import { Logger } from "tslog";

/**
 * Logging instance and settings that are shared by all scripts in this repo
 */
export const log = new Logger({
  type: "pretty",
  prettyLogTemplate: "{{hh}}:{{MM}}:{{ss}} {{logLevelName}} ",
});
