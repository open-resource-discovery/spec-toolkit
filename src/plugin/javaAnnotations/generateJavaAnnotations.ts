import * as fs from "fs";
import * as path from "path";
import { load as loadYaml } from "js-yaml";
import { log } from "../../util/log.js";
import type { JavaAnnotationsConfig } from "./configModel.js";
import type {
  SpecJsonSchema,
  JsonSchemaDefinition,
  JsonSchemaDefinitions,
} from "../../generated/spec/spec-v1/types/spec.js";

export function generateAnnotations(
  config: JavaAnnotationsConfig,
  annotatedSchemaPath: string,
  outputBase: string,
): void {
  log.info("Starting Java annotation generation");

  // Load & parse the annotated schema (JSON or YAML)
  if (!fs.existsSync(annotatedSchemaPath)) {
    log.error(`Annotated schema not found at ${annotatedSchemaPath}`);
    return;
  }
  const raw = fs.readFileSync(annotatedSchemaPath, "utf8");
  let docObj: { definitions?: JsonSchemaDefinitions };
  try {
    docObj = loadYaml(raw) as { definitions?: JsonSchemaDefinitions };
  } catch (err: unknown) {
    log.error(`Failed to parse annotated schema: ${err}`);
    return;
  }
  const defs: JsonSchemaDefinitions = docObj.definitions ?? {};
  log.info(`Loaded annotated schema with ${Object.keys(defs).length} definitions`);

  // Compute interface name from filename
  const interfaceName = path
    .basename(annotatedSchemaPath, path.extname(annotatedSchemaPath))
    .split(/[^A-Za-z0-9]/)
    .map((w) => w[0].toUpperCase() + w.slice(1))
    .join("");

  // Build output directory and filename
  const packageName = config.packageAnnotations;
  const pkgSegments = packageName.split(".");
  const outputDir = path.join(outputBase, ...pkgSegments);
  const outputFile = path.join(outputDir, `${interfaceName}.java`);

  function buildRequiredArray(def: JsonSchemaDefinition): string {
    const req = def.required ?? [];
    if (req.length === 0) return "default {}";
    return `default { ${req.map((f) => `"${f}"`).join(", ")} }`;
  }

  // resolve JSON Schema property to a Java type
  function resolveJavaType(prop: SpecJsonSchema): string {
    if (typeof prop.$ref === "string") {
      const name = prop.$ref.split("/").pop();
      return name ?? "String";
    }
    if (prop.type === "array" && prop.items) {
      // array of annotated types via anyOf
      const anyOf = Array.isArray(prop.items.anyOf) ? prop.items.anyOf : [];
      const refItem = anyOf.find((i) => typeof i.$ref === "string");
      if (refItem?.$ref) {
        return `${refItem.$ref.split("/").pop()}[]`;
      }
      // array of plain $ref
      if (typeof prop.items.$ref === "string") {
        return `${prop.items.$ref.split("/").pop()}[]`;
      }
      // fallback: primitive or nested
      return `${resolveJavaType(prop.items)}[]`;
    }
    if (prop.type === "object") return "String";
    switch (prop.type) {
      case "string":
        return "String";
      case "boolean":
        return "boolean";
      case "number":
      case "integer":
        return "double";
      default:
        return "String";
    }
  }

  function buildDefaultAnnotation(def: JsonSchemaDefinition, name: string): string {
    const req = def.required ?? [];
    const props = def.properties ?? {};
    const args = req
      .map((k) => {
        const p = props[k];
        const t = resolveJavaType(p);
        let v: string;
        if (t === "String") v = `"${typeof p.default === "string" ? p.default : ""}"`;
        else if (t === "boolean") v = `${(p.default as boolean) ?? false}`;
        else if (t === "double") v = `${(p.default as number) ?? 0}`;
        else if (t.endsWith("[]")) v = `{}`;
        else v = `@${t}()`;
        return `${k} = ${v}`;
      })
      .join(", ");
    // fully qualify nested annotation
    const qualName = `${interfaceName}.${name}`;
    return `@${qualName}${req.length ? `(${args})` : `()`}`;
  }

  // Generate methods for each property in a definition
  function generateProps(def: JsonSchemaDefinition): string {
    const props = def.properties ?? {};
    // pattern-only: single value() method
    if (!Object.keys(props).length && def.patternProperties) {
      const entry = `${Object.keys(def.patternProperties)[0]}Entry`;
      return `\n    ${entry}[] value() default {};`;
    }
    // otherwise map each property to a method signature
    return Object.entries(props)
      .filter(
        ([, s]) =>
          // skip anonymous objects without $ref/enum
          !(s.type === "object" && typeof s.$ref !== "string" && !s.enum),
      )
      .map(([k, s]) => {
        const jt = resolveJavaType(s); // Java return type
        const isRef = !!s.$ref;
        const isArr = s.type === "array";
        const anyRefs = Array.isArray(s.items?.anyOf)
          ? s.items.anyOf.filter((i): i is { $ref: string } => !!i.$ref)
          : [];
        const isAnyArr = isArr && anyRefs.length > 0;
        const isArrRef = isArr && !!s.items?.$ref;
        let dc = "";
        if (isAnyArr && anyRefs[0].$ref) {
          const rn = anyRefs[0].$ref.split("/").pop()!;
          if (defs[rn]) dc = ` default { ${buildDefaultAnnotation(defs[rn], rn)} }`;
        } else if (isArrRef && typeof s.items?.$ref === "string") {
          const rn = s.items.$ref.split("/").pop()!;
          if (defs[rn]) dc = ` default { ${buildDefaultAnnotation(defs[rn], rn)} }`;
        } else if (isRef && typeof s.$ref === "string") {
          const rn = s.$ref.split("/").pop()!;
          if (defs[rn]) dc = ` default ${buildDefaultAnnotation(defs[rn], rn)}`;
        } else if (jt === "String") {
          const v = typeof s.default === "string" ? s.default : "";
          dc = ` default "${v}"`;
        } else if (jt === "boolean") {
          dc = ` default ${(s.default as boolean) ?? false}`;
        } else if (jt === "double") {
          dc = ` default ${(s.default as number) ?? 0}`;
        } else if (isArr) {
          dc = ` default {}`;
        }
        const comment = s.description ? `    /** ${s.description.replace(/\n/g, " ")} */\n` : "";
        return `${comment}    ${jt} ${k}()${dc};`;
      })
      .join("\n\n");
  }

  // Build pattern-based Entry and wrapper interfaces
  const pattern = Object.entries(defs)
    .filter(([, d]) => !!d.patternProperties)
    .map(([n, d]) => {
      const props = d.patternProperties!;
      const firstSchema = Object.values(props)[0];
      const itemSchema = firstSchema.items ?? { type: "string" };
      const bt = resolveJavaType(itemSchema);

      return `
  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface ${n}Entry {
    String key();
    ${bt}[] values();
  }`;
    });

  const wrappers = Object.entries(defs)
    .filter(([, d]) => !!d.patternProperties)
    .map(([n, _d]) => {
      const entryName = `${n}Entry`;
      return `
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.SOURCE)
public @interface ${n} {
  ${entryName}[] value() default {};
}`;
    });

  // Build normal interfaces for definitions with properties
  const normal = Object.entries(defs)
    .filter(([, d]) => !d.patternProperties)
    .map(([n, d]) => {
      const ra = `    String[] requiredFields() ${buildRequiredArray(d)};\n`;
      const bd = Object.keys(d.properties ?? {}).length ? generateProps(d) : `\n    ${n}Entry[] value() default {};`;
      return `
  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface ${n} {
${ra}${bd}
  }`;
    });

  // concatenate all interface blocks
  const all = [...pattern, ...wrappers, ...normal].join("\n");
  const out = `package ${packageName};

import java.lang.annotation.*;

public interface ${interfaceName} {${all}

}
`;

  // Write file
  fs.mkdirSync(outputDir, { recursive: true });
  fs.writeFileSync(outputFile, out.trim(), "utf8");
  log.info(`Java annotations written to ${outputFile}`);
}
