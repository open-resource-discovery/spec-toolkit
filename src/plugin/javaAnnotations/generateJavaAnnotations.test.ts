import { jest } from "@jest/globals";
import fs from "fs-extra";
import path from "path";
import yaml from "js-yaml";
import { log } from "../../util/log.js";
import type { JavaAnnotationsConfig } from "./configModel.js";
import { generateAnnotations } from "./generateJavaAnnotations.js";

describe("generateAnnotations", () => {
  const pluginDir = path.join(process.cwd(), "src", "plugin", "javaAnnotations");
  const testDir = path.join(pluginDir, "testData");
  const outputDir = path.join(pluginDir, "tmpOutput");
  let errorSpy: ReturnType<typeof jest.spyOn>;

  beforeAll(() => {
    fs.ensureDirSync(testDir);
  });

  beforeEach(() => {
    errorSpy = jest.spyOn(log, "error").mockReturnValue(undefined);
    jest.spyOn(log, "info").mockReturnValue(undefined);
    fs.removeSync(outputDir);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  afterAll(() => {
    fs.removeSync(testDir);
    fs.removeSync(outputDir);
  });

  it("writes annotation file with default‐value cases and matches snapshot", () => {
    const schema = {
      definitions: {
        NestedAnn: {
          properties: {
            foo: { type: "string", default: "nested" },
          },
          required: ["foo"],
        },
        DefaultCases: {
          properties: {
            str: { type: "string", default: "hi" },
            bool: { type: "boolean", default: true },
            dbl: { type: "number", default: 3.14 },
            arr: { type: "array", items: { type: "integer" } },
            nested: { $ref: "#/definitions/NestedAnn" },
          },
          required: ["str", "bool", "dbl", "arr", "nested"],
        },
      },
    };

    const schemaPath = path.join(testDir, "defaults-annotated.yaml");
    fs.writeFileSync(schemaPath, yaml.dump(schema), "utf8");

    const config: JavaAnnotationsConfig = {
      packageAnnotations: "com.example.annotations",
      modelPackage: "com.example.model",
    };

    generateAnnotations(config, schemaPath, outputDir);

    const pkgDir = path.join(outputDir, ...config.packageAnnotations.split("."));
    const outFile = path.join(pkgDir, "DefaultsAnnotated.java");
    expect(fs.existsSync(pkgDir)).toBe(true);
    expect(fs.existsSync(outFile)).toBe(true);

    const content = fs.readFileSync(outFile, "utf8");
    expect(content).toMatchSnapshot();
  });

  it("logs error if annotated schema does not exist", () => {
    const missing = path.join(testDir, "nope.yaml");
    const config: JavaAnnotationsConfig = {
      packageAnnotations: "com.example.annotations",
      modelPackage: "com.example.model",
    };
    generateAnnotations(config, missing, outputDir);
    expect(errorSpy).toHaveBeenCalledWith(`Annotated schema not found at ${missing}`);
  });

  it("logs error on invalid YAML", () => {
    const badPath = path.join(testDir, "bad.yaml");
    fs.writeFileSync(badPath, "not: valid: yaml::: }", "utf8");
    const config: JavaAnnotationsConfig = {
      packageAnnotations: "com.example.annotations",
      modelPackage: "com.example.model",
    };
    generateAnnotations(config, badPath, outputDir);
    expect(errorSpy).toHaveBeenCalledWith(expect.stringContaining("Failed to parse annotated schema:"));
  });

  it("writes annotation file and matches snapshot", () => {
    const schema = {
      definitions: {
        Other: {
          properties: {},
        },
        RefTest: {
          properties: {
            r: { $ref: "#/definitions/Other" },
          },
        },
        AnyOfArr: {
          properties: {
            a: {
              type: "array",
              items: {
                anyOf: [{ $ref: "#/definitions/Other" }],
              },
            },
          },
        },
        RefArr: {
          properties: {
            b: {
              type: "array",
              items: { $ref: "#/definitions/Other" },
            },
          },
        },
        NestedArr: {
          properties: {
            c: {
              type: "array",
              items: { type: "boolean" },
            },
          },
        },
        ObjTest: {
          properties: {
            o: { type: "object" },
          },
        },
        StrTest: {
          properties: {
            s: { type: "string" },
          },
        },
        BoolTest: {
          properties: {
            t: { type: "boolean" },
          },
        },
        NumTest: {
          properties: {
            n: { type: "number" },
          },
        },
        IntTest: {
          properties: {
            i: { type: "integer" },
          },
        },
        DefaultTest: {
          properties: {
            x: { type: "unknownType" }, // default -> String
          },
        },
      },
    };
    const schemaPath = path.join(testDir, "complex-annotated.yaml");
    fs.writeFileSync(schemaPath, yaml.dump(schema), "utf8");

    const config: JavaAnnotationsConfig = {
      packageAnnotations: "com.example.annotations",
      modelPackage: "com.example.model",
    };

    generateAnnotations(config, schemaPath, outputDir);

    const pkgDir = path.join(outputDir, ...config.packageAnnotations.split("."));
    const outFile = path.join(pkgDir, "ComplexAnnotated.java");
    expect(fs.existsSync(pkgDir)).toBe(true);
    expect(fs.existsSync(outFile)).toBe(true);

    const content = fs.readFileSync(outFile, "utf8");
    expect(content).toMatchSnapshot();
  });

  it("generates pattern-only value() method when only patternProperties exist", () => {
    const schema = {
      definitions: {
        PatternOnly: {
          patternProperties: {
            pattern: { type: "string" },
          },
        },
      },
    };
    const schemaPath = path.join(testDir, "pattern-annotated.yaml");
    fs.writeFileSync(schemaPath, yaml.dump(schema), "utf8");

    const config: JavaAnnotationsConfig = {
      packageAnnotations: "com.example.annotations",
      modelPackage: "com.example.model",
    };

    generateAnnotations(config, schemaPath, outputDir);

    const pkgDir = path.join(outputDir, ...config.packageAnnotations.split("."));
    const outFile = path.join(pkgDir, "PatternAnnotated.java");
    expect(fs.existsSync(outFile)).toBe(true);

    const content = fs.readFileSync(outFile, "utf8");
    expect(content).toContain("PatternOnlyEntry[] value() default {}");
  });
});
