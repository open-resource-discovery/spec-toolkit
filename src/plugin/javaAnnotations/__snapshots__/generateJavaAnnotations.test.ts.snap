// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`generateAnnotations writes annotation file and matches snapshot 1`] = `
"package com.example.annotations;

import java.lang.annotation.*;

public interface ComplexAnnotated {
  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface Other {
    String[] requiredFields() default {};

    OtherEntry[] value() default {};
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface RefTest {
    String[] requiredFields() default {};
    Other r() default @ComplexAnnotated.Other();
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface AnyOfArr {
    String[] requiredFields() default {};
    Other[] a() default { @ComplexAnnotated.Other() };
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface RefArr {
    String[] requiredFields() default {};
    Other[] b() default { @ComplexAnnotated.Other() };
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface NestedArr {
    String[] requiredFields() default {};
    boolean[] c() default {};
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface ObjTest {
    String[] requiredFields() default {};

  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface StrTest {
    String[] requiredFields() default {};
    String s() default "";
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface BoolTest {
    String[] requiredFields() default {};
    boolean t() default false;
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface NumTest {
    String[] requiredFields() default {};
    double n() default 0;
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface IntTest {
    String[] requiredFields() default {};
    double i() default 0;
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface DefaultTest {
    String[] requiredFields() default {};
    String x() default "";
  }

}"
`;

exports[`generateAnnotations writes annotation file with default‚Äêvalue cases and matches snapshot 1`] = `
"package com.example.annotations;

import java.lang.annotation.*;

public interface DefaultsAnnotated {
  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface NestedAnn {
    String[] requiredFields() default { "foo" };
    String foo() default "nested";
  }

  @Target(ElementType.TYPE)
  @Retention(RetentionPolicy.SOURCE)
  public @interface DefaultCases {
    String[] requiredFields() default { "str", "bool", "dbl", "arr", "nested" };
    String str() default "hi";

    boolean bool() default true;

    double dbl() default 3.14;

    double[] arr() default {};

    NestedAnn nested() default @DefaultsAnnotated.NestedAnn(foo = "nested");
  }

}"
`;
